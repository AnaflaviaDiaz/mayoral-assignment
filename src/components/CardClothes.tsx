import React from "react";
import styled from "styled-components";
import Image from "next/image";

import { DiscountProps, DiscountPrice } from "./DiscountPrice";
import { ClothesProps } from "@/models/clothes.props";

const Card = styled.div`
  border: 1px solid var(--primary-color);
  flex: 0 0 50%;
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: var(--width-card-mobile);
  margin: var(--space2) var(--space1);
  border-radius: var(--border-radius-card);

  @media (min-width: 425px) {
    max-width: 200px;
  }
`;

const Figure = styled.figure`
  margin: 0px;
  width: 100%;
  padding: 0;
  display: flex;
  justify-content: center;
`;

const ImageCard = styled(Image)`
  height: auto;
  width: inherit;
  padding: var(--space1);
`;

const Title = styled.p`
  display: block;
  max-width: calc(var(--width-card-mobile) - var(--space2));
  white-space: nowrap;
  overflow: hidden !important;
  text-overflow: ellipsis;
`;

const Price = styled.span`
  color: gray;
  text-decoration: ${(props: DiscountProps) =>
    props.discountPercentage ? "line-through" : "auto"};
`;

const MoreColorsLabel = styled.p`
  color: gray;
  visibility: ${(props: MoreColorsProps) =>
    props.colors?.length ? "" : "hidden"};
`;

const AddButton = styled.button`
  background-color: var(--primary-color);
  color: white;
  cursor: pointer;
  border: 1px solid var(--primary-color);
  border-radius: 4px;
  padding: var(--space1);
  margin: var(--space1) 0 var(--space2);
  border-radius: var(--border-radius-button);
`;

interface MoreColorsProps {
  colors?: string[];
}
interface CardClothesProps extends ClothesProps {
  children?: React.ReactNode;
  alt: string;
}

export const CardClothes = ({
  title,
  alt,
  urlImage,
  price,
  discountPercentage,
  colors,
}: CardClothesProps) => {
  const getAmountWithCurrency = (price: number) => {
    const amount = price.toString().replace(".", ",");
    return `${amount} €`;
  };

  return (
    <Card>
      <Figure>
        <ImageCard src={urlImage} alt={alt} width={100} height={100} priority />
      </Figure>

      <Title>{title}</Title>

      <Price discountPercentage={discountPercentage}>
        {getAmountWithCurrency(price)}
      </Price>
      <DiscountPrice
        discountPercentage={discountPercentage || 0}
        price={price}
      />

      <MoreColorsLabel colors={colors}>más colores</MoreColorsLabel>
      <AddButton>AÑADIR</AddButton>
    </Card>
  );
};
