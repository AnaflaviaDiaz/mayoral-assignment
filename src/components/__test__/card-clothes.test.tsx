import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import { CardClothes } from "../CardClothes";

describe("CardClothes", () => {
  it("should render without discount", () => {
    const titleTest = "Name clothes test";
    render(
      <CardClothes
        title={titleTest}
        alt={titleTest}
        price={23.99}
        urlImage="https://assets.mayoral.com/images/t_auto_img/h_896/v20220420052509/12-04390-082-XL-2/jersey-con-capucha-para-nino_id_12-04390-082-M-2.jpg"
      />
    );

    const title = screen.getByText(titleTest);
    const formattedPrice = screen.getByText("23,99 €");
    expect(title).toBeInTheDocument();
    expect(formattedPrice).toBeInTheDocument();
  });

  it("should render with discount", () => {
    const titleTest = "Name clothes test";
    const { container } = render(
      <CardClothes
        title={titleTest}
        alt={titleTest}
        price={31.99}
        discountPercentage={50}
        urlImage="https://assets.mayoral.com/images/t_auto_img/h_896/v20220420052509/12-04390-082-XL-2/jersey-con-capucha-para-nino_id_12-04390-082-M-2.jpg"
      />
    );

    const title = screen.getByText(titleTest);
    const formattedPrice = screen.getByText("31,99 €");
    const formattedDiscountPriceWithPercent = screen.getByText("15,99 €(50%)");
    expect(title).toBeInTheDocument();
    expect(formattedPrice).toBeInTheDocument();
    expect(formattedDiscountPriceWithPercent).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
