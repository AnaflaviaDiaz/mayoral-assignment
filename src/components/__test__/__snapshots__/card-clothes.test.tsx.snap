// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardClothes should render with discount 1`] = `
<div>
  <div
    class="sc-17d4765f-0 jNTdvr"
  >
    <figure
      class="sc-17d4765f-1 cXlwSM"
    >
      <img
        alt="Name clothes test"
        class="sc-17d4765f-2 cFKhRt"
        data-nimg="1"
        decoding="async"
        height="100"
        src="/_next/image?url=https%3A%2F%2Fassets.mayoral.com%2Fimages%2Ft_auto_img%2Fh_896%2Fv20220420052509%2F12-04390-082-XL-2%2Fjersey-con-capucha-para-nino_id_12-04390-082-M-2.jpg&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fassets.mayoral.com%2Fimages%2Ft_auto_img%2Fh_896%2Fv20220420052509%2F12-04390-082-XL-2%2Fjersey-con-capucha-para-nino_id_12-04390-082-M-2.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fassets.mayoral.com%2Fimages%2Ft_auto_img%2Fh_896%2Fv20220420052509%2F12-04390-082-XL-2%2Fjersey-con-capucha-para-nino_id_12-04390-082-M-2.jpg&w=256&q=75 2x"
        style="color: transparent;"
        width="100"
      />
    </figure>
    <p
      class="sc-17d4765f-3 iGrPSx"
    >
      Name clothes test
    </p>
    <span
      class="sc-17d4765f-4 iECEui"
    >
      31,99 €
    </span>
    <span
      class="sc-4d5621dc-0 eNMlqQ"
    >
      15,99 €(50%)
    </span>
    <p
      class="sc-17d4765f-5 gdhSwN"
    >
      más colores
    </p>
    <button
      class="sc-17d4765f-6 bGaWor"
    >
      AÑADIR
    </button>
  </div>
</div>
`;
